### ISSUE TITLE
Spherical camera

### ISSUE THREAD

**Issue by manurare** [vispy/vispy#2292]
Hi,

Has anyone tried to render a canvas using a spherical camera with equirectangular projection? My initial idea is to create a transform: `EquirectTransform` that overrides `BaseTransform` where in `glsl_map` and `glsl_imap` you have the equirectangular projection equations which do not depend on any parameters but just the width and height of the viewbox. Something like this:
```
    glsl_map = """
        uniform int height;
        uniform int width;
        vec4 erp_transform_map(vec4 pos) {
            vec4 world2cam;
            float r = sqrt(pow(world2cam[0], 2.0) + sqrt(pow(world2cam[1], 2.0) + pow(world2cam[2], 2.0));
            float proj_xz = sqrt(pow(world2cam[0], 2.0) + sqrt(pow(world2cam[2], 2.0));
            float theta=atan(world2cam[2], world2cam[0]) + PI;
            float phi=atan(world2cam[1], proj_xz) + 0.5*PI;
            
            return vec4(width*2*theta/PI, height*phi/PI, r,1.0);
        }
    """

    glsl_imap = """
        uniform int height;
        uniform int width;
        vec4 erp_transform_imap(vec4 pos) {
            float theta = (pos[0]+0.5-width*0.5) * PI;
            float phi = (pos[1]+0.5-height*0.5) / (PI*0.5);
            float X = cos(phi) * cos(theta);
            float Y = sin(phi);
            float Z = cos(phi) * sin(theta);
            return vec4(X,Y,Z,1.0);
        }
    """
```
Then, I create a new camera class `EquirectangularCamera` which will have a `self._projection=EquirectTransform()` although I do not know how to express NDCs in equirectangular projection. I am trying to understand how the `_transform_cache` works for the `PerspectiveCamera` class in order to update the full transform in case of an event. If I understand correctly it chains transformations from pixel space to viewbox space? My problem is that I dont quite get the three transformations [here](https://github.com/vispy/vispy/blob/34d5def17a7355d7f2ac8932ff1588cd660bcf40/vispy/scene/cameras/perspective.py#L158). I am guessing `self._viewbox_tr` is the mapping from viewbox size to [-1,1], `self._projection` the NDC transformation and `camera_tr` the extrinsics? If so isn't the order between `self._projection` and `camera_tr` swapped? For me it makes sense to undo NDC first and then move the camera with  `camera_tr`.

How would you go about implementing a equirectangular renderer?

**Comment by djhoese** [comment]
So this camera would produce a cylinder essentially, right? That's at least what I think of when I hear equirectangular projection. So you could view a 2D object (ex. Image) on a cylinder on a 2D screen? Or a 3D surface (the Earth) on a 2D cylinder on a 2D screen, right?

I assume the end result/goal you're going for is that as you pan left or right you eventually wrap around to the other side of the 2D cylinder?

**Comment by manurare** [comment]
Yes! The wrap around is essential and also the rendering should be heavily distorted in the top and bottom areas

**Comment by djhoese** [comment]
I am definitely not an expert on 3D views in vispy and wrapping my head around how to implement this hurts my head a little bit. I think this is probably possible as a simple Transform object applied to the Visual. At least that's my gut instinct as I've done similar geographic projects before, but with no wrapping. My main concern would be with with discontinuity on the antimeridian. an equirectangular projection has bounds, upper/lower and left/right. However you define the view the Visuals/Canvas coordinate can't go outside of these bounds.

...right?

**Comment by manurare** [comment]
> However you define the view the Visuals/Canvas coordinate can't go outside of these bounds.

I don't quite get this. The canvas has limits of physical size in your screen but its rendering is of a sphere projected into an unfolded cylinder which I understand it has no limits? If we are talking about equirectangular rendering

Also if you are trying to render a normal image it should render only in a portion of the cylinder maybe the first 50 degrees [-25,25] both horizontally and vertically and the rest should be background color.

**Comment by djhoese** [comment]
You're not wrong. My usual use case with stuff like this is 2D projections of geographic images. The upper/lower bounds have to exist, right? For example, the Earth doesn't go beyond -90/90 latitude on an eqc projection. I suppose you just go above the original projected cylinder and go on to nothing. The left/right bounds are typically -180/180 longitude although if you are wrapping around then yeah you just keep going around.

What is a bound and what isn't is kind of odd here. Either way, do you think it requires a whole new camera? I feel like one of the existing 3D camera with a Transform on the Visuals would work...I think.

**Comment by manurare** [comment]
I think so if you go beyond latitude you would just render background color. 

Haven't take a look into the 3D cameras but I'm seeing now that they inherit from `PerspectiveCamera` which I dont need to. There is no focal length or scale factor in equirectangular projection for example. Although, It is probably overkill to create a camera from scratch as well I wanted to know how you would approach the problem. I am not familiar at all with Vispy in fact I started using just this week :\

**Comment by djhoese** [comment]
I usually let @almarklein or @rougier describe solutions when questions get to this point.

**Comment by almarklein** [comment]
> Has anyone tried to render a canvas using a spherical camera with equirectangular projection?

I don't think so :)

I think with your proposed approach you're on the right track.

> I understand correctly it chains transformations from pixel space to viewbox space? My problem is that I dont quite get the three transformations [here](https://github.com/vispy/vispy/blob/34d5def17a7355d7f2ac8932ff1588cd660bcf40/vispy/scene/cameras/perspective.py#L158). I am guessing self._viewbox_tr is the mapping from viewbox size to [-1,1], self._projection the NDC transformation and camera_tr the extrinsics?

The `camera_tr` represents the location of the camera in the scene. I had to read it a few times, but it uses `self._viewbox.scene.node_path_to_child` to get the nodes from the root of the scene to the camera (`self` is the camera), and then inverses that transform. The `viewbox_tr`is *I think* the part where the rectangular region in the canvas is selected. And `self._projection` is the camera's (perspective) projection. This is the part you will want to replace.

I'm not sure if `self._projection` maps to NDC, or whether its logical coordinates or something (in the latter case `viewbox_tr` would include the transform to NDC)

**Comment by manurare** [comment]
Thanks for the answer. Unfortunately I think it is not as trivial as it seems. If I modify the `self._projection` to implement a shader that applies trigonometric functions instead of a 4x4 matrix and two things happen:

1) Triangles on the antimeridian (where the cylinder wraps) could have vertices at -179.9° and +179.9° which translates to vertices at -0.999 and 0.999 (ignore exact decimal precision)  resulting in a drawn triangle which spans the whole viewbox. 
2) Triangles do not curve when reaching the poles. They are still drawn as straight lines.

I saw in forums that the correct way is to project the sphere into a cubemap and then render 6 perspective cameras (one per face) and stitch the result into an equirectangular projection. While this fixes 1 it is not clear to me how it fixes 2. 

----
Another question about the perspective camera. Why does the near and far clipping planes are derived using [this](https://github.com/vispy/vispy/blob/bf793b36010d3cecb8307e1360f6549633fed1e5/vispy/scene/cameras/perspective.py#L167)? Isn't the near clipping plane directly the focal length and the far plane any arbitrary value bigger than the focal length?

**Comment by almarklein** [comment]
> Unfortunately I think it is not as trivial as it seems.

I wasn't assuming this to be trivial :)

I'm not sure how to handle (1). The straight lines could be handled by using a denser grid. They would still be straight, but if the triangles are short enough, it would matter less.

Using a cubemap might be a good solution that (I think) might solve both problems. I don't know if anyone has ever done something with a cubemap in vispy though, so it might be a bit of a bumpy road too.


> Another question about the perspective camera. Why does the near and far clipping planes are derived using [this](https://github.com/vispy/vispy/blob/bf793b36010d3cecb8307e1360f6549633fed1e5/vispy/scene/cameras/perspective.py#L167)? Isn't the near clipping plane directly the focal length and the far plane any arbitrary value bigger than the focal length?

No, the best values for the near and far plane depend on the situation. I can't remember this particular formula, but I suppose it's a trick to get planes that are generally good.

**Comment by manurare** [comment]
I will try first in openGL and then I will try to add an equirectangular camera to vispy I think it is a good option to have in the library!

**Comment by os-gabe** [comment]
Hey @manurare,
Are you basically wanting to make a 360 viewer with vispy? I did this by making a custom PanoVisual which is a sphere with the equirectangular image texture mapped onto it. You then have a lot of flexibility with what you do with your camera. You can place the camera outside the sphere and it's like you are looking at a globe, or you can place your camera inside the sphere and you get the normal 360 image viewer experience.
![Peek 2022-08-11 20-57](https://user-images.githubusercontent.com/47365564/184282659-a9191c7a-f96a-4c10-a672-a3e0fda3aab9.gif)

**Comment by kmuehlbauer** [comment]
@os-gabe That looks great!

**Comment by manurare** [comment]
Hey @os-gabe . Thanks for that example, it looks amazing! 

I have managed to implement in the end in openGL. I have a depthmap for my spherical image so mapping it to unit sphere is not suitable for my task. I am looking to generate some parallax when I translate the camera. If an sphere is used as geometry, then there is no parallax whereas if you apply the depthmap as the geometry you can feel parallax when moving around. Also I want the rendering to be an equirectangular image, in the viewer you made I can see the rendering is using a pinhole camera. 

Nice job!
